File: python/day11.py
#!/usr/bin/env python3
import sys
from collections import defaultdict

def parse(lines):
    g = defaultdict(list)
    for line in lines:
        line = line.strip()
        if not line:
            continue
        name, rhs = line.split(":", 1)
        name = name.strip()
        targets = [t.strip() for t in rhs.strip().split() if t.strip()]
        g[name] = targets
    return g

# ---------- Part 1 ----------
def count_paths(g, start="you", goal="out"):
    memo = {}
    onstack = set()
    def dfs(u):
        if u == goal:
            return 1
        if u in memo:
            return memo[u]
        if u in onstack:
            return 0
        onstack.add(u)
        total = 0
        for v in g.get(u, []):
            total += dfs(v)
        onstack.remove(u)
        memo[u] = total
        return total
    return dfs(start)

# ---------- Part 2 ----------
def count_paths_with_both(g, start="svr", goal="out", a="dac", b="fft"):
    memo = {}
    onstack = set()
    def dfs(u, va, vb):
        if u == goal:
            return 1 if va and vb else 0
        key = (u, va, vb)
        if key in memo:
            return memo[key]
        if key in onstack:
            return 0
        onstack.add(key)
        va2 = va or (u == a)
        vb2 = vb or (u == b)
        total = 0
        for v in g.get(u, []):
            total += dfs(v, va2, vb2)
        onstack.remove(key)
        memo[key] = total
        return total
    return dfs(start, False, False)

def main():
    if len(sys.argv) != 2:
        print("Usage: reactor.py input.txt")
        sys.exit(1)
    with open(sys.argv[1]) as f:
        lines = f.readlines()
    g = parse(lines)

    part1 = count_paths(g, "you", "out")
    print(f"Part 1: paths you->out = {part1}")

    total_paths = count_paths(g, "svr", "out")
    constrained = count_paths_with_both(g, "svr", "out", "dac", "fft")
    print(f"Part 2: total svr->out = {total_paths}")
    print(f"Part 2: paths visiting dac & fft = {constrained}")

if __name__ == "__main__":
    main()

File: perl/day11.pl
#!/usr/bin/perl
use strict;
use warnings;

sub parse_graph {
    my ($lines) = @_;
    my %g;
    for my $line (@$lines) {
        next unless $line =~ /\S/;
        my ($name, $rhs) = split /:/, $line, 2;
        $name =~ s/^\s+|\s+$//g;
        $rhs  =~ s/^\s+|\s+$//g;
        my @targets = grep { $_ ne '' } split /\s+/, $rhs;
        $g{$name} = \@targets;
    }
    return \%g;
}

# ---------- Part 1 ----------
sub count_paths {
    my ($g, $start, $goal) = @_;
    my %memo;
    my %onstack;
    my $dfs;
    $dfs = sub {
        my ($u) = @_;
        return 1 if $u eq $goal;
        return $memo{$u} if exists $memo{$u};
        return 0 if $onstack{$u};
        $onstack{$u} = 1;
        my $total = 0;
        for my $v (@{ $g->{$u} // [] }) {
            $total += $dfs->($v);
        }
        delete $onstack{$u};
        $memo{$u} = $total;
        return $total;
    };
    return $dfs->($start);
}

# ---------- Part 2 ----------
sub count_paths_with_both {
    my ($g, $start, $goal, $a, $b) = @_;
    my %memo;
    my %onstack;
    my $dfs;
    $dfs = sub {
        my ($u, $va, $vb) = @_;
        if ($u eq $goal) {
            return ($va && $vb) ? 1 : 0;
        }
        my $key = join("|", $u, $va ? 1 : 0, $vb ? 1 : 0);
        return $memo{$key} if exists $memo{$key};
        return 0 if $onstack{$key};
        $onstack{$key} = 1;
        my $va2 = $va || ($u eq $a);
        my $vb2 = $vb || ($u eq $b);
        my $total = 0;
        for my $v (@{ $g->{$u} // [] }) {
            $total += $dfs->($v, $va2, $vb2);
        }
        delete $onstack{$key};
        $memo{$key} = $total;
        return $total;
    };
    return $dfs->($start, 0, 0);
}

sub main {
    @ARGV == 1 or die "Usage: reactor.pl input.txt\n";
    my $file = shift @ARGV;
    open my $fh, '<', $file or die "Can't open $file: $!";
    my @lines = <$fh>;
    close $fh;

    my $g = parse_graph(\@lines);

    my $part1 = count_paths($g, 'you', 'out');
    print "Part 1: paths you->out = $part1\n";

    my $total = count_paths($g, 'svr', 'out');
    my $both  = count_paths_with_both($g, 'svr', 'out', 'dac', 'fft');
    print "Part 2: total svr->out = $total\n";
    print "Part 2: paths visiting dac & fft = $both\n";
}

main();
